### ISSUE TITLE
Why not Pydantic, dataclasses, attrs, grpc, ctypes Structures, etc.?

### ISSUE THREAD

**Issue by ConnorSMaynes** [rogervila/py_dto#4]
: )

This package is the first result on DDG, so I feel it worth asking.

Why use this over any of the options mentioned in the title?

**Comment by rogervila** [comment]
Hi @ConnorSMaynes :)

`py_dto` focuses on the following topics:
- **Type validation**: if a field does not match the excepted type, a `TypeError` is raised
- **Key validation**: if a value is passed for a key that does not exist or a key is missing a value (_wip right now!_), a `ValueError` is raised

**Pydantic** is a great tool that includes data conversion, custom validators, and other features that are out of the scope of DTOs.

**Ctypes** behave differently on  Linux, Windows, and Mac OS X, which can lead to unexpected issues.

**gRPC** might be a little bit overwhelming for small projects or prototypes.

Python native **dataclasses** do not have type validation, which is the main feature of this package. Example:

```py
from dataclasses import dataclass

@dataclass
class User:
    name: str

u = User(name=123)

print(u.name) # 123
print(type(u.name)) # <class 'int'>
```

So, in conclusion, the other suggested tools have different abilities than `py_dto`, and this is the reason for the existence of this package.

I hope this answer helps!

**Comment by rogervila** [comment]
Closing for now, if you have any other questions do not hesitate to open a new issue :)
